### PR TITLE
:computer: Added support for initial volume calculation 

### PR DESCRIPTION
**Pull request by cw646**
Closed PR #46 and reopened here as branch fell behind Develop.

### CODE REVIEW

**Review by kks32 — CHANGES_REQUESTED** [review]
Could we have CATCH unit test

File: include/element.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
+
+
+//! Calculate volume of an element
+//! \tparam Tdim Dimension of the mesh
+//! \tparam Tvertices Number of vertices in an element
+//! \param[in] 3x8 Matrix of element vertices coordinates
```
Comment (kks32) [inline include/element.cc]: There are no input parameters

File: include/element.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      |  /    |  /
+      | /     | /
+      |/      |/
+      7-------6
+
+      * Calculation of hexahedron volume from
+      * https://arc.aiaa.org/doi/pdf/10.2514/3.9013
+   */
+
+  Eigen::Vector3d a(coordinates_.at(3)[0], coordinates_.at(3)[1],
```
Comment (kks32) [inline include/element.cc]: Make these `const`s.

File: include/element.h [hunk 3]
```diff
@@ -61,4 +61,52 @@ inline Eigen::VectorXd shapefn(const std::array<double, 2>& xi) {
 }
 };
 
+//! Element class to hold element vertices id and coordinates
+//! \param[in] element id, vertices id, vertices coordinates
+class Element {
+
+ public:
+  //! constructor for Element class
+  //! \param[in] element id and vertices id's
```
Comment (kks32) [inline include/element.h:72]: Should be on separate lines

File: include/element.h [hunk 4]
```diff
@@ -61,4 +61,52 @@ inline Eigen::VectorXd shapefn(const std::array<double, 2>& xi) {
...
+  Element(unsigned id, const Eigen::VectorXd& vertices) : id_{id} {
+
+    vertices_ = vertices;
+  }
+
+  //! Return volume of element
+  double calculate_volume();
+
+  //!\param[in] coordinates of all element vertices
```
Comment (kks32) [inline include/element.h]: Define what function does

File: include/gmsh.h [hunk 5]
```diff
@@ -55,11 +55,11 @@ class GMSH : public Mesh<Tdim, Tvertices> {
   //! Map to store element id and vertices id
   using Mesh<Tdim, Tvertices>::elements_;
 
-  //! Map to store element id and vertices coordinates
-  using Mesh<Tdim, Tvertices>::elementcoordinates_;
-
   //! Vector of material points
   using Mesh<Tdim, Tvertices>::materialpoints_;
+
+  //! Map of volumes
```
Comment (kks32) [inline include/gmsh.h]: Do we need this still, can we not just call volume?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/element.cc (hunk 1 above)
Comment (kks32) [reply include/element.cc]: `\retval` for return

File: include/element.cc [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+                    coordinates_.at(3)[2]),
+      b(coordinates_.at(2)[0], coordinates_.at(2)[1], coordinates_.at(2)[2]),
+      c(coordinates_.at(1)[0], coordinates_.at(1)[1], coordinates_.at(1)[2]),
+      d(coordinates_.at(0)[0], coordinates_.at(0)[1], coordinates_.at(0)[2]),
+      e(coordinates_.at(7)[0], coordinates_.at(7)[1], coordinates_.at(7)[2]),
+      f(coordinates_.at(6)[0], coordinates_.at(6)[1], coordinates_.at(6)[2]),
+      g(coordinates_.at(5)[0], coordinates_.at(5)[1], coordinates_.at(5)[2]),
+      h(coordinates_.at(4)[0], coordinates_.at(4)[1], coordinates_.at(4)[2]);
+
+  volume =
```
Comment (kks32) [inline include/element.cc]: const volume

File: include/gmsh.tcc [hunk 7]
```diff
@@ -171,23 +171,25 @@ template <unsigned Tdim, unsigned Tvertices>
...
   //! Iterate through element_
   for (const auto& element : elements_) {
 
+    std::vector<Eigen::Vector3d> verticescoordinates;
+
     //! Iterate through the vertices to get coordinates depending on the element
     for (unsigned j = 0; j < Tvertices; ++j) {
       //! Get the vertex wanted from the id
-      auto verticesfind = vertices_.find(element.second[j]);
+
```
Comment (kks32) [inline include/gmsh.tcc]: Explain in comments

File: include/io.h [hunk 8]
```diff
@@ -43,6 +44,12 @@ class IO {
   //! Return mesh file name
   std::string mesh_file_name() const { return mesh_filename_; }
 
+  //! Create output file names
+  boost::filesystem::path output_file(const std::string& attribute,
+                                      const std::string& file_extension);
+
+  void write_volumes(const std::map<unsigned, double>& volumes);
```
Comment (kks32) [inline include/io.h]: Comments, input

File: include/io.h [hunk 9]
```diff
@@ -57,10 +64,13 @@ class IO {
   std::string mesh_filename_;
 
   //! File name of vertices
```
Comment (kks32) [inline include/io.h]: file name for material points

File: include/mesh.h [hunk 10]
```diff
@@ -57,14 +63,17 @@ class Mesh {
...
+  std::vector<std::unique_ptr<Element>> elements_;
 
   //! Map to store element id and vertices coordinates
   std::map<unsigned, Eigen::VectorXd> elementcoordinates_;
 
   //! Vector of material points
   std::vector<std::unique_ptr<MaterialPoints<Tdim>>> materialpoints_;
+
+  //! Map to store element id and volume
+  std::map<unsigned, double> volumes_;
```
Comment (kks32) [inline include/mesh.h]: Remove this

File: include/mesh.tcc [hunk 11]
```diff
@@ -54,4 +54,27 @@ void Mesh<Tdim, Tvertices>::compute_stresses() {
...
+//! Compute initial volume of material points
+//! \tparam Tdim Dimension of the mesh
+//! \tparam Tvertices Number of vertices in an element
+template <unsigned Tdim, unsigned Tvertices>
+void Mesh<Tdim, Tvertices>::compute_volumes() {
+
+  unsigned arrayposition = 0;
+
+  double volume = 0;
+  std::array<double, Tdim> volumearray;
```
Comment (kks32) [inline include/mesh.tcc]: What is volumearray

File: include/mesh.tcc [hunk 12]
```diff
@@ -54,4 +54,27 @@ void Mesh<Tdim, Tvertices>::compute_stresses() {
...
+
+  for (const auto& element : elements_) {
+
+    unsigned id = element->id();
+
+    volumes_.insert(std::make_pair<unsigned, double>(
```
Comment (kks32) [inline include/mesh.tcc]: No need to store

File: tests/volume_test.cc [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  SECTION("Check volume of hexahedron") {
+
+    std::vector<Eigen::Vector3d> coordinates;
+    Eigen::VectorXd vertices(8);
+
+    vertices << 0, 0, 0, 0, 0, 0, 0, 0;
+
+    Element element(id, vertices);
+
+    Eigen::Vector3d a(0, 0.1, 0.1), b(0.05, 0.1, 0.1), c(0.05, 0.05, 0.1),
```
Comment (kks32) [inline tests/volume_test.cc]: const

Comment (kks32) [inline tests/volume_test.cc]: `{}` to initialize?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/element.cc (hunk 1 above)
Comment (ezrayst) [reply include/element.cc]: do you want to make it `.tcc` btw? or want to keep it as `cc`?

Comment (kks32) [reply include/element.cc]: This is not a templatised class, so it won't be a `.tcc` file.

**Review by kks32 — APPROVED** [review]

**Pull request merged**